### ISSUE TITLE
How to trap and report Errors

### ISSUE THREAD

**Issue by DPineault** [calcinai/xero-php#471]
I'm trying to create Invoices and in a general sense I have it functional, but now I need to harden the code by handling errors that may arise and I can't seem to understand how to accomplish this.

My basic code (there is more but not truly relevant to the current question - just adding Line Items, ...)

            $invoice = new Invoice($xero);
            $invoice->setType(INVOICE::INVOICE_TYPE_ACCREC);
            $invoice->setStatus(INVOICE::INVOICE_STATUS_SUBMITTED);
            $invoice->setContact($contact);
            $invoice->setInvoiceNumber($inv->invNo);
            //Add, Dates, ...,  Line Items
            $XeroInvoiceResponse = $invoice->save();

I've seen a few mentions of using try { but no real concrete examples that I could implement.  The wiki example has nothing regarding this either.  Are there any examples or documentation that I could refer to?

For a concrete example, I am trying to capture when the user tries to submit the same invoice more than once and Xero returns "Invoice # must be unique." validation error.

Thank you for any help you can provide.

**Comment by timacdonald** [comment]
Hey @DPineault I've just created a Pull Request to add documentation about this, as you are correct - there is no mention of error handling.

Until the PR gets merged into this repo (if it is accepted) you can look at my branch to see the documentation I've added: https://github.com/timacdonald/xero-php/tree/error_docs#handling-errors

**Comment by DPineault** [comment]
Thank you for the reply.

I have tried, but still can't seem to get it working?

           ```
 try{
                $invoice->save();
            } catch (BadRequestException $exception) {
                echo "BadRequestException";
                exit();
            } catch (InternalErrorException $exception) {   
                echo "InternalErrorException";
                exit();
            } catch (NotAvailableException $exception) {   
                echo "NotAvailableException";
                exit();
            } catch (NotFoundException $exception) {   
                echo "NotFoundException";
                exit();
            } catch (NotImplementedException $exception) {   
                echo "NotImplementedException";
                exit();
            } catch (OrganisationOfflineException $exception) {   
                echo "OrganisationOfflineException";
                exit();
            } catch (RateLimitExceededException $exception) {   
                echo "RateLimitExceededException";
                exit();
            } catch (UnauthorizedException $exception) {   
                echo "UnauthorizedException";
                exit();
            }
```

**Comment by mogilvie** [comment]
Correct, you need either a use statement to allow the class shorthand to work, or a fully qualified class name in the catch.

**Comment by DPineault** [comment]
Any chance you could give me an example to learn from (still very new to this)?

**Comment by DPineault** [comment]
Thank you.  It works!

For anyone else that is trying to do the same below is a very simple example.
```
try {
    $xero->save($invoice);
} catch (BadRequestException $exception) {
    echo $exception->getMessage();
    die();
}
```
and obviously you can build upon that for all the different exceptions.

**Comment by timacdonald** [comment]
Glad to hear @DPineault 

You can close this issue if you think you got it sorted.

**Comment by DPineault** [comment]
May I further this discussion to delve a little deeper into the Bad Request exception.  [The documentation states](https://developer.xero.com/documentation/api/http-response-codes)

> HTTP 400 responses include an “ApiException" element in the response that contains a useful summary of the reason for the error.
> 
> {
>   "ErrorNumber": 10,
>   "Type": "ValidationException",
>   "Message": "A validation exception occurred",
>   "Elements": [{
>     "ValidationErrors": [{
>       "Message": "Email address must be valid"
>     }] 
>   }] 
> }

Right now in my catch I have
`echo $exception->getMessage();`
How can I get the Type and ValidationErrors details?

**Comment by mogilvie** [comment]
The payloads of the exception can be different depending on the exception class and the API that you're querying.
You can use var_dump($exception) to view the exception structure and content.
Often the responses from an API are json encoded. So you need to find the method that returns the data you want, then json_decode() to convert the string to an array. Use regular array manipulation to get the details.
